### PR TITLE
Added Docker File

### PR DESCRIPTION
**Pull request by Abhinavks1405**

### CODE REVIEW

File: Dockerfile
Comment (SaashaJoshi) [inline Dockerfile:1]: Can this be edited to support python versions from 3.8 to 3.12?

File: Dockerfile
Comment (SaashaJoshi) [inline Dockerfile:4]: I am unsure if `venv` needs to be initiated when creating a docker image. Do you have a particular reason behind doing so?

**Review by SaashaJoshi — CHANGES_REQUESTED** [review]
Hi @Abhinavks1405, thanks for the input. I just left a few reviews and the rest LGTM.

**Comment by SaashaJoshi** [comment]
Also, do have a look at the errors while lining the file. Let me know if you will need any help there!

File: Dockerfile
Comment (Abhinavks1405) [reply Dockerfile:4]: Having virtual env in docker ensures that there is consistency accors different environment and not have a package conflict

**Comment by Abhinavks1405** [comment]
@SaashaJo

**Comment by Abhinavks1405** [comment]
@SaashaJoshi 
I have commited the desired changes.
1. added support for python version 3.8 to 3.12.
2. updated pylint-> v3.1.0 to remove the error occurred while linting with python 3.12 .
3. added build.yml to have github action workflows .

File: .github/workflows/build.yml
Comment (SaashaJoshi) [inline .github/workflows/build.yml:3]: I would prefer if Docker images were made only on a workflow dispatch.

```suggestion
on: 
    workflow_dispatch:
```

File: .github/workflows/build.yml
Comment (SaashaJoshi) [inline .github/workflows/build.yml:1]: Could you rename the file to Build Docker? Also `build_docker.yml"

File: .github/workflows/build.yml
Comment (SaashaJoshi) [inline .github/workflows/build.yml:13]: You can now use `checkout@v3`

File: .github/workflows/build.yml
Comment (SaashaJoshi) [inline .github/workflows/build.yml:17]: I think `your-app-name` is a placeholder that should probably be replaced with `piqture`

Comment (SaashaJoshi) [inline .github/workflows/build.yml:17]: Also, `piqture` app versions are different from python versions. You will have to add another strategy matrix that takes into account the app version.

File: .github/workflows/build.yml
Comment (SaashaJoshi) [inline .github/workflows/build.yml:10]:
```suggestion
        python-version: ['3.8', '3.9', '3.10', '3.11', '3.12']
        app_veriosn: ["v0.1.1"]
```

File: .github/workflows/build.yml
Comment (SaashaJoshi) [inline .github/workflows/build.yml:21]: The `your-app` placeholder needs to be changed here as well.

File: .github/workflows/unit_test.yml
Comment (SaashaJoshi) [inline .github/workflows/unit_test.yml:43]: This is not the correct place to have a GitHub token. You will have to provide me with the token and I will add it as a secret on GitHub.

Comment (SaashaJoshi) [reply .github/workflows/unit_test.yml:43]: Which token are we trying to implement here?

File: .pre-commit-config.yaml
Comment (SaashaJoshi) [inline .pre-commit-config.yaml:34]: This is an unrelated change for the PR. You will have to start another PR to make this change.

File: Dockerfile
Comment (SaashaJoshi) [reply Dockerfile:8]: But Docker itself creates separate containers for this purpose.

**Review by SaashaJoshi — CHANGES_REQUESTED** [review]

File: Dockerfile
Comment (SaashaJoshi) [inline Dockerfile:1]: Can we have a default value of 3.9

File: Dockerfile
Comment (Abhinavks1405) [reply Dockerfile:8]: Understood , i thought we are working with multiple python version in the same docker , which is not he case

File: .github/workflows/unit_test.yml
Comment (Abhinavks1405) [reply .github/workflows/unit_test.yml:43]: My bad , I didn't noticed that github tokens got exposed.

Comment (Abhinavks1405) [reply .github/workflows/unit_test.yml]: the github token is automatically set and it is required because without it the upload coverage to coverall is failing

**Comment by Abhinavks1405** [comment]
i have checked and done all the possible changes , the only issue currently facing and previously mentioned is of github tokens .

**Comment by Abhinavks1405** [comment]
@SaashaJoshi  , waiting for you review on pr .

**Comment by SaashaJoshi** [comment]
Hey, can you revert all the changes made to `unit-test.yml` and `.pre-commit.yml`? These files weren't meant to be changed.

**Comment by Abhinavks1405** [comment]
@SaashaJoshi reverted changes in unit_test.yml and .pre-commit-config.yaml

File: .github/workflows/build.yml
Comment (SaashaJoshi) [inline .github/workflows/build.yml:12]: We do not need outputs from our yml file. Can we safely remove this?

**Review by SaashaJoshi — CHANGES_REQUESTED** [review]

**Comment by SaashaJoshi** [comment]
There is only one thing that I notice here. We are building Docker images twice for each version and each app version. However, what we wish to do is create Docker images for `piQturev0.1.1` with each python version. Let me suggest some changes.

File: .github/workflows/build.yml
Comment (SaashaJoshi) [inline .github/workflows/build.yml:23]:
```suggestion
        docker build --build-arg PYTHON_VERSION=${{ matrix.python-version }} \
                     --build-arg APP_VERSION=${{matrix.app-version}} \
                     -t piqture:${{ matrix.app-version }}-${{ matrix.python-version }} .
```

Comment (SaashaJoshi) [reply .github/workflows/build.yml:23]: With these changes, we will not require the `build-app` yml workflow.

File: .github/workflows/build.yml
Comment (SaashaJoshi) [inline .github/workflows/build.yml:20]:
```suggestion
    - name: Build Docker image for App and Python versions
```

File: .github/workflows/build.yml
Comment (SaashaJoshi) [inline .github/workflows/build.yml:49]: Any of this will not be needed with the above suggested changes.

File: .github/workflows/build.yml
Comment (SaashaJoshi) [inline .github/workflows/build.yml:13]: We do not need this variable. Instead define `app-version` here.
```suggestion
      app-version: ["0.1.1"]
```

**Review by SaashaJoshi — COMMENTED** [review]
Suggested changes to reduce the Docker workflow to build app and python versions of the app once only. Let me know what you think!

**Comment by Abhinavks1405** [comment]
> There is only one thing that I notice here. We are building Docker images twice for each version and each app version. However, what we wish to do is create Docker images for `piQturev0.1.1` with each python version. Let me suggest some changes.

the point which I processed was mainly scalability as having 2 matrices to handle python version and app version so that we don't have to continuously change logic each time , mainly having 2 docker build ensures that I can have multiple app version running on same python version image reducing build time . i maily proceeded with this logic but want to know your thoughts as they are helpful .

**Comment by SaashaJoshi** [comment]
I agree with your point, But doesn't it mean we will have, let's say 5 Docker versions per app version? This is also what I am trying to achieve with a single Docker build. For example, the build will produce `piQture-1.1-3.8` then `piQture-1.1-3.9`, and so on.

**Comment by Abhinavks1405** [comment]
> I agree with your point, But doesn't it mean we will have, let's say 5 Docker versions per app version? This is also what I am trying to achieve with a single Docker build. For example, the build will produce `piQture-1.1-3.8` then `piQture-1.1-3.9`, and so on.

Ok I understand your point , I can try to do this with tags to bring it to a single docker build otherwise the app version can be hardcoded .

**Comment by SaashaJoshi** [comment]
Hi @Abhinavks1405, any updates? I will try to get this PR merged by today. You can then open a PR to make corrective measures after the hacktoberfest marked PR is merged.

**Comment by Abhinavks1405** [comment]
> Hi @Abhinavks1405, any updates? I will try to get this PR merged by today. You can then open a PR to take corrective measures after the hacktoberfest marked PR is merged.

yes yes the changes are made , let me push it

**Comment by Abhinavks1405** [comment]
@SaashaJoshi, can I raise 2 pr after this, one related to the updated Pylint version for supporting python 3.12 and other one for docker file optimizations ?

**Comment by SaashaJoshi** [comment]
Sure, feel free to create as many PRs. But just make sure that each PR serves a single purpose.

**Pull request merged**
